web/src/main/java/org/openmrs/web/filter/update/UpdateFilter.java

```diff
@@ -334,54 +334,116 @@ protected boolean authenticateAsSuperUser(String usernameOrSystemId, String pass
 			connection = DatabaseUpdater.getConnection();
 			
 			String select = "select user_id, password, salt from users where (username = ? or system_id = ?) and retired = '0'";
-			PreparedStatement statement = connection.prepareStatement(select);
-			statement.setString(1, usernameOrSystemId);
-			statement.setString(2, usernameOrSystemId);
-			
-			if (statement.execute()) {
-				ResultSet results = statement.getResultSet();
-				if (results.next()) {
-					Integer userId = results.getInt(1);
-					DatabaseUpdater.setAuthenticatedUserId(userId);
-					String storedPassword = results.getString(2);
-					String salt = results.getString(3);
-					String passwordToHash = password + salt;
-					boolean result = Security.hashMatches(storedPassword, passwordToHash) && isSuperUser(connection, userId);
-					statement.close();
-					return result;
+			PreparedStatement statement = null;
+			try {
+				statement = connection.prepareStatement(select);
+				statement.setString(1, usernameOrSystemId);
+				statement.setString(2, usernameOrSystemId);
+
+				if (statement.execute()) {
+					ResultSet results = null;
+					try {
+						results = statement.getResultSet();
+						if (results.next()) {
+							Integer userId = results.getInt(1);
+							DatabaseUpdater.setAuthenticatedUserId(userId);
+							String storedPassword = results.getString(2);
+							String salt = results.getString(3);
+							String passwordToHash = password + salt;
+							boolean result = Security.hashMatches(storedPassword, passwordToHash)
+							        && isSuperUser(connection, userId);
+							connection.close();
+							results.close();
```

Comment:
Please close connection. results and statements in a finally block. If one of close operation cause error it will stop closing other statements. 

try {
} catch() {
} finally {
//close statements going here
}